### PR TITLE
Fix cursor appears in front of digit when highlighting amount and entering digit 

### PR DESCRIPTION
**Pull request by bernhardoj**
### Details

The selection is updated manually and from the onSelectionChange event which could gives a wrong value.

### Fixed Issues

$ https://github.com/Expensify/App/issues/41762
PROPOSAL: https://github.com/Expensify/App/issues/41762#issuecomment-2104107478

### Tests

Same as QA Steps
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as QA Steps
### QA Steps

1. Open a split bill page
2. Enter any amount
3. Select any user
4. Select all digits on any participant split amount input and replace it with any number
5. Verify the selection is at the right of the amount
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/50919443/28f8d92c-3961-4adf-ac97-10b7ef436c99

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/50919443/fcc0914a-49ba-47d0-88b9-a19b092f439e

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/50919443/6215d756-e391-4802-a7bb-96b8377734e8

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/50919443/ed888d72-3ffa-43ca-92f4-c2a15620de16

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/50919443/9cdbce34-0773-4458-9c4c-d243ab6f2314

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/50919443/593dcb75-d083-4cd7-bbea-66c99f96fdb8

</details>

### CODE REVIEW

**Comment by jayeshmangwani** [comment]
@bernhardoj I am getting some inconsistencies between staging and this PR on the web. If we try to double-click on the input value, it selects all the values on staging, but in our PR, double-clicking does nothing. Please confirm whether you are getting the same results.

Also, when I try to press between the input values on the web, selection doesn't go to that value. e.g. if input values is 789, we press on 1st index then cursor should be at the 8.

https://github.com/Expensify/App/assets/35371050/7a9138be-fe38-45b3-b223-3a747017005c

**Comment by bernhardoj** [comment]
@jayeshmangwani I can reproduce it in `main` too. Most likely happen after it's migrated to selection list here.
https://github.com/Expensify/App/pull/40785

**Comment by jayeshmangwani** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/35371050/f984b762-d0ad-4137-9137-fb56fc3cc790




</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/35371050/37d0adc5-fcfe-4583-893f-132e2bc97add




</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/35371050/56da7554-f4a2-4a9b-9393-d6cbe38dc043




</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/35371050/37dc0016-701b-465d-bd30-86b7ad71914d




</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/35371050/2b26b3fe-0365-4491-a05b-5f918d6ddd71




</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/35371050/c669437e-e865-498b-9c9e-a34e855b5fc8




</details>

**Review by jayeshmangwani — APPROVED** [review]
LGTM 🚀

**Comment by youssef-lr** [comment]
I'm using the adhoc build https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/42158/index.html

**Comment by bernhardoj** [comment]
Hmm, I can't reproduce it. I have merged with main, idk if that helps.

https://github.com/Expensify/App/assets/50919443/f4cbc15a-f54f-404c-aece-d9443823a119

Btw, looks like the split calculation is broken. I enter the amount as 25 but both participant split is also 25 instead of 12.5

**Comment by youssef-lr** [comment]
>Btw, looks like the split calculation is broken. I enter the amount as 25 but both participant split is also 25 instead of 12.5 can't reproduce it anymore, idk what step I took that cause this issue before I can repro it again on web, really weird

I'm not able to reproduce this one

**Comment by youssef-lr** [comment]
>Hmm, I can't reproduce it. I have merged with main, idk if that helps.

You have to put the cursor in the beginning of the amount, not the end, so in your recording try to tap next to the left of the digit `2`

**Comment by bernhardoj** [comment]
Oh, I see you put it on the end in your video, so I follow it 😂, but I am still unable to reproduce it.

https://github.com/Expensify/App/assets/50919443/f30c43d7-b025-4feb-853e-f75701bd2263

**Comment by jayeshmangwani** [comment]
I am also not able to reproduce this issue.

https://github.com/Expensify/App/assets/35371050/f985a604-7122-426e-b421-ed10daecb094

**Comment by youssef-lr** [comment]
Testing again shortly

**Comment by youssef-lr** [comment]
I'm still able to reproduce :/ I tap exactly to the left of the first digit

https://github.com/Expensify/App/assets/9680864/19e640f6-953c-4c2e-b4c2-88b468ed7620

**Comment by youssef-lr** [comment]
Or you can also tap on the currency

**Comment by bernhardoj** [comment]
Hmm, still can't reproduce it. In your video, when you tap the currency, the selection moves to the right, that's why when you press delete, the rightmost digit is removed. Maybe it's reproducible in physical devices only? I don't have a physical iOS device.

Btw, have you checked whether it's reproducible in `main` too?

**Comment by jayeshmangwani** [comment]
I have tested the Android ad-hoc, and the issue is not happening for me; my device is not added to iOS testing, so I can not confirm in iOS adhoc

**Comment by youssef-lr** [comment]
>Hmm, still can't reproduce it. In your video, when you tap the currency, the selection moves to the right, that's why when you press delete, the rightmost digit is removed

Yes that works as expected, but the selection should remain at the end, no?

**Comment by bernhardoj** [comment]
Yeah, I would expect it to remain at the end too, but I think the real problem here is this one,

> when you tap the currency, the selection moves to the right,

My guess is that the `selection` state of the component
https://github.com/Expensify/App/blob/e8ae3c5acedf0e6788dc574c7b6f3043ca37092a/src/components/MoneyRequestAmountInput.tsx#L120-L123

doesn't match the TextInput internal native `selection`. So, the (MoneyRequestAmountInput)`selection` state is still at the left (correct one), but the TextInput internal `selection` is at the right (wrong one), and when we update the value, it makes sense for the cursor to "move" to the left because the `prevSelection` state is at the left.

https://github.com/Expensify/App/blob/e8ae3c5acedf0e6788dc574c7b6f3043ca37092a/src/components/MoneyRequestAmountInput.tsx#L154

Maybe it's related to this comment
https://github.com/Expensify/App/blob/e8ae3c5acedf0e6788dc574c7b6f3043ca37092a/src/components/Composer/index.native.tsx#L27-L28

I think we should just apply this fix for Android? Or all except iOS?

**Comment by youssef-lr** [comment]
Making another build to test

**Comment by youssef-lr** [comment]
@bernhardoj can you please merge with main?

**Comment by bernhardoj** [comment]
Merged with main

**Comment by jayeshmangwani** [comment]
@youssef-lr Is it possible to add my iOS device for ad-hoc testing builds? I can reproduce the original issue on my iOS device in app store build. To verify this PR ad-hoc, my UDID needs to be added for testing

**Comment by jayeshmangwani** [comment]
>  Is it possible to add my iOS device for ad-hoc testing builds? I can reproduce the original issue on my iOS device in app store build. To verify this PR ad-hoc, my UDID needs to be added for testing

@youssef-lr Tested in latest app store build and issue is not happening for me now(it was happening for me on the previous version), can you please check on your side if issue is still happening for you ?

**Comment by youssef-lr** [comment]
>Is it possible to add my iOS device for ad-hoc testing builds? I

@jayeshmangwani I'm not familiar with how this is done, but I'll ask for you.

>Tested in latest app store build and issue is not happening for me now(it was happening for me on the previous version), can you please check on your side if issue is still happening for you ?

I just reproduced it in the latest build. I'm making a build right now of this PR to test again.

**Comment by youssef-lr** [comment]
>I think we should just apply this fix for Android? Or all except iOS?

Can we try it in this PR? I'm still reproducing the issue on latest build

**Comment by bernhardoj** [comment]
Updated to only apply it on android

**Comment by youssef-lr** [comment]
Yeah that fixed it!

**Review by youssef-lr — APPROVED** [review]

File: src/libs/actions/Report.ts [hunk 1]
```diff
@@ -1027,7 +1027,6 @@ function navigateToAndOpenReportWithAccountIDs(participantAccountIDs: number[])
  */
 function navigateToAndOpenChildReport(childReportID = '0', parentReportAction: Partial<ReportAction> = {}, parentReportID = '0') {
     if (childReportID !== '0') {
-        openReport(childReportID);
```
Comment (jayeshmangwani) [inline src/libs/actions/Report.ts:1030]: @bernhardoj I am confused why we removed the `openReport` call from here when I have approved the PR then this change was not here?

Comment (bernhardoj) [reply src/libs/actions/Report.ts:1030]: Ah, my bad. It's from my lint commit here: https://github.com/Expensify/App/pull/42158/commits/6d0630d35122717a484469b25e78fad35dbaa825

I think I accidentally committed it when working on this [PR](https://github.com/Expensify/App/pull/43250)

Comment (jayeshmangwani) [reply src/libs/actions/Report.ts:1030]: Oh, it is from another PR, but it was intentional to remove it, so there is no need to revert this change, right?

Comment (bernhardoj) [reply src/libs/actions/Report.ts:1030]: Yes, it was intentional to remove it.

**Pull request merged**